### PR TITLE
Modify collection's `Debug` output to resemble in their content only

### PR DESCRIPTION
**Pull request by tbu-**
r? @Gankro

### CODE REVIEW

**Comment by tbu-** [comment]
reem voiced some concerns on IRC (2015-02-10 22:00:00 +0100):

    [22:16:45] <reem> eibwen: I thought the decision was to keep types in Debug output
    [22:17:19] <eibwen> oh, was it?
    [22:17:24] <eibwen> I'll dig up the RFC
    [22:17:32] <eibwen> I thought it was the other way around
    [22:18:15] <eibwen> reem: from the RFC: Focus on the *runtime* aspects of a type; repeating information such as suffixes for integer literals is not generally useful since that data is readily available from the type definition.
    [22:19:03] <eibwen> reem: further down: Containers print using *some* notation that makes their type and contents clear. (Since we lack literals for all container types, this will be ad hoc).
    [22:19:17] <eibwen> this is indicated by using [] for sequences and {} for sets/maps
    [22:19:30] <reem> eibwen: [] or {} is not enough to tell me the tyep
    [22:19:37] <reem> if I want to know the type I need a name too
    [22:20:15] <eibwen> I might well have misunderstood the RFC, but I interpret it as "focus on the semantics of the type, not its implementation"
    [22:20:49] <eibwen> which is: "sequence" for Vec, DList, etc. (the actual type can be seen in source code, the semantic type is there)
    [22:21:00] <eibwen> or "map" for HashMap, TreeMap, etc.
    [22:21:55] <bluss_> I like it. I think it composes better with derived Debug impls of custom types? (But those still use struct names(?))
    [22:22:25] <reem> I still think the types should stay in the output
    [22:22:47] <eibwen> I think the terser the output, the better
    [22:22:52] <eibwen> as long as it's readable
    [22:23:02] <eibwen> because more things fit on the screen
    [22:23:25] <eibwen> (this has been an actual nuisance for me, if you nest a few structs, the debug output explodes)

**Comment by Gankro** [comment]
Conventions issue, punting to r? @aturon

File: src/libcollections/bit.rs [hunk 1]
```diff
@@ -977,7 +977,7 @@ impl Ord for Bitv {
 impl fmt::Debug for Bitv {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
         for bit in self {
-            try!(write!(fmt, "{}", if bit { 1u32 } else { 0u32 }));
+            try!(write!(fmt, "{}", if bit { 1 } else { 0 }));
```
Comment (Gankro) [inline src/libcollections/bit.rs]: This is an interesting case. A big binary string seems reasonable, I suppose.

Comment (tbu-) [inline src/libcollections/bit.rs]: Didn't change anything here though.

Comment (Gankro) [inline src/libcollections/bit.rs]: Yeah totally. Poking around it just got me thinking.

**Comment by Gankro** [comment]
Seems to be missing fixes for BTree* if you meant to be exhaustive.

Also apparently BinaryHeap has *no* impl??

**Comment by tbu-** [comment]
@Gankro Implemented for `BTree*` as well.

**Comment by alexcrichton** [comment]
Citing [RFC 565](https://github.com/rust-lang/rfcs/blob/master/text/0565-show-string-guidelines.md#debugging-fmtdebug)

> Containers print using some notation that makes their type and contents clear. (Since we lack literals for all container types, this will be ad hoc).

(just pointing out that this is currently intended)

**Comment by tbu-** [comment]
@alexcrichton I understand that as: Containers print something that makes their semantic type clear (`[]` for sequence, e.g. `Vec`, `DList`, etc., `{}` for maps, sets, like `HashSet`, `HashMap`, ...).

**Comment by alexcrichton** [comment]
I would personally be ok with that interpretation, and I'm in favor of this change.

**Comment by aturon** [comment]
@tbu- Thanks for this PR!

I'm also in favor of these changes and interpretation of the RFC, particularly since none of the current outputs can be used to reconstruct the collection.

(Aside: if we had macros for constructing each collection, you could imagine printing `vec![1, 2]` and `hash_map["foo": 1, "bar": 2]` etc.)

cc @seanmonstar @wycats

**Comment by tbu-** [comment]
@aturon Any update? Should I rebase this?

**Comment by aturon** [comment]
@tbu- As I said [in an earlier comment](https://github.com/rust-lang/rust/pull/22157#issuecomment-73928215), I'm happy with these changes -- sorry nothing has happened since then!

If you rebase, please ping @alexcrichton since I am away on vacation this week.

**Comment by alexcrichton** [comment]
Yes I'd be fine merging with a rebase

**Comment by tbu-** [comment]
Rebased.

**Comment by Manishearth** [comment]
(Not a real failure, happened because I forced the rollup through)

**Comment by Manishearth** [comment]
@bors: retry

(not a real failure)

**Comment by Manishearth** [comment]
@bors: retry

(Please ignore the failures for now)

**Pull request merged**